docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+            {
+                sourceContext.AddSource("additionalFiles.cs", @"
+namespace Generated
+{
+    public class AdditionalTextList
+    {
+        public static void PrintTexts()
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}
+");
+            });
+
+            // register the output context.RegisterOutput(output); }); } }
+```
+
+## Advanced Implementation
+
+### Combining and filtering
+
+While the transformation steps outlined above allow a user to create simple
+generators from a single source of data, in reality it is expected that an
+author will need a way to take multiple sources of data and combine them.
+
+There exists an extension method `Combine` that allows the author to take one
+data source and merge it with another, for example, extracting a set of types
+from the compilation and using them to generate something on a per additional
+file basis.
+
+```csharp
+public static partial class IncrementalValueSourceExtensions
+{
+    // join 1 => many ((source1[0], source2), (source1[0], source2), ...)
+    public static IncrementalValueSource<(T, IEnumerable<U>)> Combine<T, U>(this IncrementalValueSource<T> source1, IncrementalValueSource<U> source2) => ...
+}
+```
+
+The second data source is batched, and the resulting step has a value type of
+`(T, IEnumerable<U>)`. That is, a tuple where each element consists of a single
+item of data from `source1` combined with the entire batch of data from
+`source2`. While this is somewhat low-level, when combined with subsequent
+transforms, it gives the author the ability to combine an arbitrary number of
+data sources into any shape they require.
+
+In the following example the author combines the additional text source with the
+compilation:
+
+```csharp
+var combined = context.Sources.AdditionalTexts.Combine(context.Sources.Compilation);
+```
+
+The type of combined is an
+`IncrementalValueSource<(AdditionalText, IEnumerable<Compilation>)>`. For each
+additional text, there is a tuple with the text, and the batched data of the
+compilation source. As the author knows that there is only ever a single
+compilation, they are free to transform the data to select the single
+compilation object:
+
+```csharp
+IncrementalValueSource<(AdditionalText, Compilation)> transformed = combined.Transform(pair => (pair.Item1, pair.Item2.Single()));
+```
+
+Similarly a cross join can be achieved by first combining two value sources,
+then batch transforming the resulting value source.
+
+```csharp
+IncrementalValueSource<(AdditionalText, MetadataReference)> combined = context.Sources.AdditionalTexts
+                                                                                      .Combine(context.Sources.MetadataReference)
+                                                                                      .TransformMany(pair => pair.Item2.Select(item2 => (pair.Item1, item2));
```

Comment:
fwiw, this became compeltely impossible for me to reason about.  :)   It might be worthwhile to hide behind a helper.  So  `.AdditionalTexts.CrossJoin(context.Sources.MetadataReferences)`.